### PR TITLE
Cody eval TypeScript dataset

### PR DESCRIPTION
**Pull request by taras-yemets**
Iteration 1/n on adding TypeScript dataset to Cody completion evaluation suite.

Screenshot 2023-10-13 at 12 25 05

## Test plan
This PR adds completion evaluation datasets.

### CODE REVIEW

File: vscode/test/benchmark/datasets/typescript/infers-type-open-file/config.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "entryFile": "generate.ts",
+  "openFiles": ["generate.ts", "types.ts"],
+  "closedFiles": [],
+  "solutionFile": "solution.ts",
+  "testFile": "test.ts",
+  "testCommand": "ts-node"
```
Comment (umpox) [inline vscode/test/benchmark/datasets/typescript/infers-type-open-file/config.json:7]: @taras-yemets Just an FYI, I've updated the main branch to require the file is provided explicitly for this (as you may not always want a `testFile`, like @slimsag mentioned for Go)

```suggestion
  "testCommand": "ts-node test.ts"
```

**Comment by taras-yemets** [comment]
@umpox, @slimsag, what do you think about the test cases added? Do they test real-world scenarios in your opinion? I'm looking for feedback on the overall direction.

File: vscode/test/benchmark/datasets/typescript/external-package-lodash/generate.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import lodash from 'lodash'
+
+export function areSame(obj1: any, obj2: any): boolean {
+    return █
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/external-package-lodash/generate.ts:4]: I'm concerned if packages like `lodash` are good examples here. It's easier to write test cases for them as many devs are familiar with them, but LLMs know a lot of them too, and may just guess (or remember) the correct answers.

**Comment by taras-yemets** [comment]
@philipp-spiess, tagging you here too as you're actively contributing to the quality evaluation discussion.

File: vscode/test/benchmark/datasets/typescript/external-package-deep-equal/config.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "entryFile": "generate.ts",
+  "openFiles": [],
+  "closedFiles": ["package.json"],
+  "solutionFile": "solution.ts",
+  "testFile": "test.ts",
+  "testCommand": "pnpm install && ts-node test.ts"
```
Comment (umpox) [inline vscode/test/benchmark/datasets/typescript/external-package-deep-equal/config.json:7]: The `install` in these tests is going to be quite slow. I think it's fine right now, but we might want to consider just moving these to the https://github.com/sourcegraph/cody-evaluation-datasets repo and committing node_modules there if we find ourselves adding more of these

**Review by umpox — APPROVED** [review]
Nice!

**Comment by taras-yemets** [comment]
Re: https://github.com/sourcegraph/cody/pull/1332#issuecomment-1761109676

@umpox, we got significantly better results for Copilot for this dataset too.

cc: @philipp-spiess

File: vscode/test/benchmark/datasets/typescript/external-package-deep-equal/config.json (hunk 3 above)
Comment (taras-yemets) [reply vscode/test/benchmark/datasets/typescript/external-package-deep-equal/config.json:7]: It's slow indeed!
We have only two test cases requiring the installation of dependencies, so we can leave it here for now. But as the dataset grows I agree we should put it into [cody-evaluation-datasets](https://github.com/sourcegraph/cody-evaluation-datasets) repo.

File: vscode/test/benchmark/datasets/typescript/class-instance-method-same-file/generate.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+export class Rectangle {
+    constructor(
+        private _height: number,
+        private _width: number,
+        private _borderWidth: number
+    ) {}
+
+    get area(): number {
+        return █
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/class-instance-method-same-file/generate.ts:9]: |Cody|Copilot|
|--|--|
|Screenshot 2023-10-13 at 13 01 45|Screenshot 2023-10-13 at 13 01 51|

☝🏻 These results are are pretty stable across runs.

Comment (philipp-spiess) [reply vscode/test/benchmark/datasets/typescript/class-instance-method-same-file/generate.ts:9]: @taras-yemets Can you check the request log for the Cody autocomplete request? Maybe our `suffix` size is too short and does not include the `calcArea` method below?

Comment (taras-yemets) [reply vscode/test/benchmark/datasets/typescript/class-instance-method-same-file/generate.ts:9]: Logs include `calcArea`
```
{
        "speaker": "human",
        "text": "Below is the code from file path generate.ts. Review the code outside the XML tags to detect the functionality, formats, style, patterns, and logics in use. Then, use what you detect and reuse methods/libraries to complete and enclose completed code only inside XML tags precisely without duplicating existing implementations. Here is the code: \n```\nexport class Rectangle {\n    constructor(\n        private _height: number,\n        private _width: number,\n        private _borderWidth: number\n    ) {}\n\n<CODE5711>get area(): number {\n        return</CODE5711>\n    }\n\n    get width(): number {\n        return this._width + this._borderWidth * 2\n    }\n\n    get height(): number {\n        return this._height + this._borderWidth * 2\n    }\n\n    calcArea(): number {\n        return this.width * this.height\n    }\n}\n\n```"
      },
      ```

File: vscode/test/benchmark/datasets/typescript/external-package-deep-equal-dynamic-import/generate.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+export async function areSame(obj1: any, obj2: any): Promise<boolean> {
+    const isEqual = await import█
+    return isEqual(obj1, obj2)
+}
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply vscode/test/benchmark/datasets/typescript/external-package-deep-equal-dynamic-import/generate.ts:4]: |Cody|Copilot|
|--|--|
|Screenshot 2023-10-13 at 14 26 39|Screenshot 2023-10-13 at 14 26 47|

Both fail to complete correctly:
- Cody suggests an existing package name (`lodash.isequal`) unlike Copilot (`lodash/isEqual` doesn't exist on npm)
- Both ignore `package.json` content - `deep-equal` package is listed there as prokect dependency

File: vscode/test/benchmark/datasets/typescript/function-argument-type/generate.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+interface Person {
+    name: string
+    age: █
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/function-argument-type/generate.ts:3]: |Cody|Copilot|
|--|--|
|Screenshot 2023-10-13 at 14 33 02|Screenshot 2023-10-13 at 14 33 10|

Copilot usually correctly completes `age` type (`string`) based on the `createPerson` implementation. Cody seems to "guess" the type based on the field name (I would expect age to be a number too if we didn't have the implementation).

File: vscode/test/benchmark/datasets/typescript/node-write-file/generate.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import fs from 'fs/promises'
+
+export async function writeToFile(path: string, content: string): Promise<void> {
+    try {
+        █
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/node-write-file/generate.ts:5]: Cody usually returns better completions for this case. Copilot doesn't come up with any completion from time to time.

File: vscode/test/benchmark/datasets/typescript/object-creation-closed-file/generate.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import { Person } from './types'
+
+
+export function createPerson(fullName: string, yearsOld: number): Person {
+    return {█
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/object-creation-closed-file/generate.ts:5]: |Cody|Copilot|
|--|--|
|Screenshot 2023-10-13 at 14 40 36|Screenshot 2023-10-13 at 14 40 44|

When `types.ts` is not currently open in the editor, Cody usually tries to guess the return value and returns incorrect results. Copilot, in this case, often doesn't suggest anything.
When `type.ts` is open in the editor or interface `Person` is defined in the same file (`generate.ts`), Cody usually suggests the correct completion (including the `createdAt` field). Copilot often doesn't suggest anything.

**Comment by taras-yemets** [comment]
> @taras-yemets Have you tested the two cases where the suffix should have a bigger impact (`calcArea` and `createPerson`) with StarCoder? I feel like maybe the fake infilling we have with Claude Instant might be the culprit here but 😬

@philipp-spiess, I'm running the benchmark command `BENCHMARK_DISABLE_FEATURE_FLAGS=true` against the dotcom instance, so consider these results to be for the default completions provider.

**Comment by umpox** [comment]
@taras-yemets Can experiment with Starcoder/different models by updating the file at `vscode/test/benchmark/fixtures/workspace/.vscode/settings.json`

with:
```
  "cody.autocomplete.advanced.provider": "unstable-fireworks",
  "cody.autocomplete.advanced.model": "starcoder-16b",
```

File: vscode/test/benchmark/datasets/typescript/object-property-access-closed-file/generate.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+import { Person } from './types'
+
+export function getAge(person: Person): number {
+    const currentYear = new Date().getFullYear()
+    return currentYear - person.█
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/object-property-access-closed-file/generate.ts:5]: |Cody|Copilot|
|--|--|
|Screenshot 2023-10-13 at 14 55 36|Screenshot 2023-10-13 at 14 55 43|

Unlike https://github.com/sourcegraph/cody/pull/1334/files#r1358155970, for this case, Copilot relatively stable provides correct suggestion. Note that `types.ts` is not open in the editor in this case. I don't think that it's a guess game as `birthYear` completion seems more obvious than `yearOfBirth`.

Comment (taras-yemets) [reply vscode/test/benchmark/datasets/typescript/object-property-access-closed-file/generate.ts:5]: Both provide correct suggestions when `types.ts` is open.

File: vscode/test/benchmark/datasets/typescript/promise-closed-file/generate.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import { fetchCurrentUser, fetchUser } from './api'
+import { Person } from './types'
+
+export async function getUser(): Promise<Person> {
+    const person = █
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/promise-closed-file/generate.ts:5]: Results similar to https://github.com/sourcegraph/cody/pull/1334/files#r1358166188

File: vscode/test/benchmark/datasets/typescript/utils-closed-file/generate.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+import { Club } from "./types";
+import { isRich, totalBudget } from './utils'
+
+export function richClubsTotalBudget(clubs: Club[]): number {
+    return █
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/utils-closed-file/generate.ts:5]: Results similar to https://github.com/sourcegraph/cody/pull/1334/files#r1358166188

File: vscode/test/benchmark/datasets/typescript/utils-import-all-closed-file/generate.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+import { Club } from "./types";
+import * as utils from './utils'
+
+export function richClubsTotalBudget(clubs: Club[]): number {
+    return █
```
Comment (taras-yemets) [inline vscode/test/benchmark/datasets/typescript/utils-import-all-closed-file/generate.ts:5]: Both Cody and Copilot fail to suggest correct completions when using `import * as utils from './utils'` and `utils.ts` is closed. Cody usually provides correct suggestion when the file is open, while Copilot seems to fail in this case too.

**Comment by taras-yemets** [comment]
> @taras-yemets Can experiment with Starcoder/different models by updating the file at `vscode/test/benchmark/fixtures/workspace/.vscode/settings.json`
> 
> with:
> 
> ```
>   "cody.autocomplete.advanced.provider": "unstable-fireworks",
>   "cody.autocomplete.advanced.model": "starcoder-16b",
> ```

@philipp-spiess, got the same result for https://github.com/sourcegraph/cody/pull/1334#discussion_r1358060118 with the suggested model.

**Comment by philipp-spiess** [comment]
@taras-yemets Let's chat next week about this in more detail. I wonder if we can get something going before the release so we can A/B test this. This is great work!

**Comment by philipp-spiess** [comment]
@umpox happy to add you too if you want :)

**Comment by philipp-spiess** [comment]
@taras-yemets oh also make sure to share this in Slack this is super awesome insight! :)

**Comment by umpox** [comment]
@philipp-spiess Yes please would be good to discuss :)

**Pull request merged**